### ISSUE TITLE
BUG: A pm.Deterministic is added to the model.unobserved_RVs (as well as the model.deterministics)

### ISSUE THREAD

**Issue by jonsedar** [pymc-devs/pymc#6869]
### Describe the issue:

I ran into this while creating a model that has a skeleton (see `mdlb` in my MRE below) that uses a Deterministic.

### Issue: 

The `pm.Deterministic` is added to the `model.unobserved_RVs`, as well as the `model.deterministics`. 

### Impact

I cant say for sure that this is the wrong thing to do, but it does seem illogical, and does mean that:

1. **`mdlb.str_repr()` fails** because it doesn't expect to have to `str_repr` a DeterministicRV

Note the **`yhat`** in `print(mdlb.unobserved_RVs)` in the output below shouldn't be there IMO

```
# mdlb
# [b, sigma, **yhat**]  # <-- why is this yhat here in the unobservedRVs?
# []
# [pot-yhat]
# [yhat]  # <-- I do indeed expect to see yhat here in the deterministics
```

### Reproduceable code example:

```python
import numpy as np
import pymc as pm

with pm.Model() as mdla:
    # 0. Create data objects for obs (X,Y)
    y = pm.MutableData('y', np.random.normal(loc=3, scale=0.1, size=100))
    
    # # 1. Create linear model
    b = pm.Normal('b', mu=0.0, sigma=1.0)
    sigma = pm.InverseGamma('sigma', alpha=5.0, beta=4.0)

    # 2. Condition using Observed
    _ = pm.Normal('yhat', mu=b, sigma=sigma, observed=y)

    # 3. Forward estimation already supported without any more effort

print('mdla')
print(mdla.unobserved_RVs)
print(mdla.observed_RVs)
print(mdla.potentials)
print(mdla.deterministics)


with pm.Model() as mdlb:
    # 0. Create data objects for obs (X,Y)
    y = pm.MutableData('y', np.random.normal(loc=3, scale=0.1, size=100))
    
    # # 1. Create linear model
    b = pm.Normal('b', mu=0.0, sigma=1.0)
    sigma = pm.InverseGamma('sigma', alpha=5.0, beta=4.0)

    # # 2. Condition using Potential
    norm_rv = pm.Normal.dist(mu=b, sigma=sigma)
    _ = pm.Potential('pot-yhat', pm.logp(norm_rv, y))

    # # 3. Within-model forward estimation to still provide 
    # # sample_prior_predictive and sample_posterior_predictive and overcome: 
    # # UserWarning: The effect of Potentials on other parameters is ignored during prior predictive sampling.
    yhat = pm.Deterministic('yhat', norm_rv)


print('\nmdlb')
print(mdlb.unobserved_RVs)
print(mdlb.observed_RVs)
print(mdlb.potentials)
print(mdlb.deterministics)


# NOTE: output printed
# mdla
# [b, sigma]
# [yhat]
# []
# []

# mdlb
# [b, sigma, yhat]
# []
# [pot-yhat]
# [yhat]


print(mdla.str_repr())

# works fine


print(mdlb.str_repr())
```

### Error message:

```shell
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
Cell In[96], line 1
----> 1 print(mdlb.str_repr())

File ~/.mamba/envs/oreum_lab/lib/python3.10/site-packages/pymc/printing.py:103, in str_for_model(model, formatting, include_params)
     99 """Make a human-readable string representation of Model, listing all random variables
    100 and their distributions, optionally including parameter values."""
    101 all_rv = itertools.chain(model.unobserved_RVs, model.observed_RVs, model.potentials)
--> 103 rv_reprs = [rv.str_repr(formatting=formatting, include_params=include_params) for rv in all_rv]
    104 rv_reprs = [rv_repr for rv_repr in rv_reprs if "TransformedDistribution()" not in rv_repr]
    106 if not rv_reprs:

File ~/.mamba/envs/oreum_lab/lib/python3.10/site-packages/pymc/printing.py:103, in (.0)
     99 """Make a human-readable string representation of Model, listing all random variables
    100 and their distributions, optionally including parameter values."""
    101 all_rv = itertools.chain(model.unobserved_RVs, model.observed_RVs, model.potentials)
--> 103 rv_reprs = [rv.str_repr(formatting=formatting, include_params=include_params) for rv in all_rv]
    104 rv_reprs = [rv_repr for rv_repr in rv_reprs if "TransformedDistribution()" not in rv_repr]
    106 if not rv_reprs:

File ~/.mamba/envs/oreum_lab/lib/python3.10/site-packages/pymc/printing.py:150, in str_for_potential_or_deterministic(var, formatting, include_params, dist_name)
    148 else:  # plain
    149     if include_params:
--> 150         return rf"{print_name} ~ {dist_name}({_str_for_expression(var, formatting=formatting)})"
...
    231     )
    232 else:
--> 233     return r"f(" + ", ".join([n.strip("$") for n in names]) + ")"

AttributeError: 'NoneType' object has no attribute 'strip'
```

### PyMC version information:

Python implementation: CPython
Python version       : 3.10.12
IPython version      : 8.14.0

pymc    : 5.6.1
pytensor: 2.12.3

Compiler    : Clang 15.0.7 
OS          : Darwin
Release     : 22.6.0
Machine     : x86_64
Processor   : i386
CPU cores   : 8
Architecture: 64bit

scipy     : 1.11.1
matplotlib: 3.7.2
numpy     : 1.24.3
pytensor  : 2.12.3
pymc      : 5.6.1
arviz     : 0.16.1
pandas    : 2.0.3
seaborn   : 0.12.2

**Comment by ricardoV94** [comment]
The name is perhaps not the best, but you may be interested in `model.free_RVs` instead of `model.unobserved_RVs`. 

The latter are all the quantities that should show up in your trace after prior sampling. If you include `observed_RVs`, you have all the variables that should show up in the trace after prior_predictive sampling.

The corresponding `model.unobserved_value_vars` include all the variables that should show up in the trace after `pm.sample` (the same as `model.unobserved_RVs` plus untransformed projections of value variables):

https://github.com/pymc-devs/pymc/blob/f8581ab84ace71c3a82e3b8eed569c63af84208d/pymc/model.py#L926-L934

https://github.com/pymc-devs/pymc/blob/f8581ab84ace71c3a82e3b8eed569c63af84208d/pymc/model.py#L865-L885

The error you are getting is unrelated and due to https://github.com/pymc-devs/pymc/issues/6311
In this case your `norm_rv` doesn't have this `str_repr` monkey patched method and the model fails. We really should get rid of that assumption!

**Comment by jonsedar** [comment]
Gotcha, thanks @ricardoV94 for such a complete answer! I'll close this one then, and wait for a fix to #6311

Would a clumsy temporary fix for me be to make a child class with a monkey-patched `str_repr`?

**Comment by ricardoV94** [comment]
> Gotcha, thanks @ricardoV94 for such a complete answer! I'll close this one then, and wait for a fix to #6311
> 
> Would a clumsy temporary fix for me be to make a child class with a monkey-patched `str_repr`?

An equally easy temporary fix is not not look for `str_repr` in the model repr and instead use the single function that we are monkey patching instead. That easy fix would be the perfect fix for the issue btw ;)
